### DISCUSSION TITLE
mutate form relationship  data

### DISCUSSION THREAD

**Discussion by topegret** [filamentphp/filament#4133] [Help] [answered]
https://filamentphp.com/docs/2.x/admin/resources/creating-records#customizing-data-before-saving

per document we can mutate form data before save into database table.  I noticed that if I  dd($data) .  only the direct model data in the array but no relationship data. 
say i have a scenario that a form scehma to create a 'user' together with one to one relationship 'profile'.    

```
    ...
return $form
        ->schema([             
            Group::make()
            ->schema([
                Card::make()->schema([
                    Fieldset::make(label:'Agent Account')->schema([
                        TextInput::make('name'),
                         TextInput::make('email'),
                                ])->columns(1),
                        ... ...
                            ]),  
                     ]) ->columnSpan(['lg' => 1]),


            Forms\Components\Group::make()->relationship('profile')
            ->schema([
                Card::make()->schema([
                    Fieldset::make(label:'Contact')->schema([
                                 TextInput::make('mobile'),
                                   TextInput::make('phone'),
                        .......
                           ])
                      ])
                      .....
```

in above sample code. when dd,  data array has name and email. but not any of the relationship data such as 'mobile' and 'phone'  . 
 Is there any way to obtain and then mutate the relationship data  before saving?  thanks!

**Comment by topegret** [comment]
OK. it seems I figured out the alternative. what I can do is to make changes in the afterCreate() function in CreateUser class. 
```
protected function afterCreate(): void
    {
         $profile = $this->record->profile;
           //make change of profile items  then save 
         $profile->save();
    } 
```

**Comment by danharrin** [comment]
Created a feature request to track this - #4176

**Reply by spiritkiddie** [reply]
@danharrin would it be possible to pass all the form data including any relationship data to the mutatedatabeforesave(array $data). In my case I'm trying to generate e.g serial number for the main form based on a relationship field value but at this point I'm having difficulties in doing that. Or is there any better way to handle such a scenario.

**Reply by danharrin** [reply]
Pass `->dehydrated()` to the relationship layout component

**Comment by danharrin** [comment, marked as answer]
Hey @topegret! This has been fixed in #4424 - you can now use `mutateRelationshipDataBeforeCreateUsing()` and `mutateRelationshipDataBeforeSaveUsing()` on the Group!

```php
->mutateRelationshipDataBeforeSaveUsing(fn (array $data) => ...)
```

**Reply by topegret** [reply]
Great Thanks!

**Reply by MhdTarhini** [reply]
Hello, there a way to handle record relationship data inside the record create or edit class ?

I'm tying to get record relationship data before update them

**Reply by leandrocfe** [reply]
if you add `->dehydrated()` in the relationship, you can save it manually

**Reply by MhdTarhini** [reply]
Thanks!

**Reply by hasan-aftab** [reply]
Brother @danharrin what If wanna mutate relationship data before saving in db without group?

**Reply by bhattji** [reply]
Where do you define mutateRelationshipDataBeforeSaveUsing() ?

I have a relationmanager where I want to use it? how?

**Comment by mkwsra** [comment]
Just in case if you reached this page while looking for "how to mutate Relation Manager's form" then consider overriding this function within your Relation Manager

```php
protected function configureCreateAction(Tables\Actions\CreateAction $action): void
    {
        parent::configureCreateAction($action);
        $action->mutateFormDataUsing(function ($data) {
            // Example:
            unset($data['foo']);
            
            return $data;
        });

    }
```

**Reply by danharrin** [reply]
You don't need to, you can chain it on to `CreateAction::make()` in the `$table->headerActions()`

**Reply by mkwsra** [reply]
Thanks for your comment, but where can I do so when I'm using a relation manager? let's say I have User who has AddressesRelationManager and I would like to mutate the data of the address form, in this case where can I use the `$table` variable

**Reply by danharrin** [reply]
<img width="722" alt="Screenshot 2024-07-01 at 15 21 01" src="https://github.com/filamentphp/filament/assets/41773797/f1da3764-5d88-4a3c-b037-4f9d5002f3fb">

**Reply by mkwsra** [reply]
Thank you so much!! I was so much confused, and it's all my bad! your help is much appreciated ❤️

**Comment by NaungYeHtet** [comment]
How about using CheckBoxList with BelongsToMany relationship?

**Comment by PegeDev** [comment]
i use mutateRelationshipDataBeforeFillUsing in group layout form, and i change with mutate .. the data not filled in update. how to resolve issue .. 

![image](https://github.com/user-attachments/assets/5e6d3ac2-69b5-4d60-b085-efbb5f165049)
![image](https://github.com/user-attachments/assets/442ead56-cdb0-4ee4-b4b0-5e804e5c96dc)
